File: problems/289_game-of-life/main.cpp
#include <iostream>
#include <sstream>
#include <vector>

using namespace std;

class Solution
{
public:
    void gameOfLife(vector<vector<int>> &board)
    {
        int m = (int)board.size();
        int n = (int)board[0].size();

        static const int dr[8] = {-1, -1, -1, 0, 0, 1, 1, 1};
        static const int dc[8] = {-1, 0, 1, -1, 1, -1, 0, 1};

        // 1st pass: mark transitions using 2 and 3
        for (int r = 0; r < m; ++r)
        {
            for (int c = 0; c < n; ++c)
            {
                int live = 0;
                for (int k = 0; k < 8; ++k)
                {
                    int nr = r + dr[k], nc = c + dc[k];
                    if (nr < 0 || nr >= m || nc < 0 || nc >= n)
                        continue;
                    int v = board[nr][nc];
                    if (v == 1 || v == 2)
                        live++; // currently alive
                }

                if (board[r][c] == 1)
                {
                    if (live < 2 || live > 3)
                        board[r][c] = 2; // live -> dead
                }
                else
                { // dead
                    if (live == 3)
                        board[r][c] = 3; // dead -> live
                }
            }
        }

        // 2nd pass: finalize to 0/1
        for (int r = 0; r < m; ++r)
        {
            for (int c = 0; c < n; ++c)
            {
                board[r][c] &= 1; // 0->0, 1->1, 2->0, 3->1
            }
        }
    }
};

string board_to_string(const vector<vector<int>> &board)
{
    stringstream ss;
    ss << "[";
    for (size_t i = 0; i < board.size(); ++i)
    {
        if (i > 0)
        {
            ss << ", ";
        }
        ss << "[";
        for (size_t j = 0; j < board[i].size(); ++j)
        {
            if (j > 0)
            {
                ss << ", ";
            }
            ss << board[i][j];
        }
        ss << "]";
    }
    ss << "]";
    return ss.str();
}

int main()
{
    Solution sol;

    vector<vector<vector<int>>> samples = {
        {{0, 1, 0}, {0, 0, 1}, {1, 1, 1}, {0, 0, 0}},
        {{1, 1}, {1, 0}}};

    vector<vector<vector<int>>> expected = {
        {{0, 0, 0}, {1, 0, 1}, {0, 1, 1}, {0, 1, 0}},
        {{1, 1}, {1, 1}}};

    for (size_t i = 0; i < samples.size(); ++i)
    {
        vector<vector<int>> board = samples[i];
        sol.gameOfLife(board);

        cout << "Sample " << (i + 1) << ": board = " << board_to_string(samples[i]) << "\n";
        cout << "Placeholder result: " << board_to_string(board) << "\n";
        cout << "Expected output: " << board_to_string(expected[i]) << "\n\n";
    }

    return 0;
}

File: problems/289_game-of-life/main.rs
struct Solution;

impl Solution {
    pub fn game_of_life(board: &mut Vec<Vec<i32>>) {
        if board.is_empty() || board[0].is_empty() {
            return;
        }
        let m = board.len();
        let n = board[0].len();
        let dr = [-1, -1, -1, 0, 0, 1, 1, 1];
        let dc = [-1, 0, 1, -1, 1, -1, 0, 1];

        // 1st pass: mark transitions
        for r in 0..m {
            for c in 0..n {
                let mut live = 0;
                for k in 0..8 {
                    let nr = r as i32 + dr[k];
                    let nc = c as i32 + dc[k];
                    if nr >= 0 && nr < m as i32 && nc >= 0 && nc < n as i32 {
                        let v = board[nr as usize][nc as usize];
                        if v == 1 || v == 2 {
                            live += 1;
                        }
                    }
                }
                if board[r][c] == 1 {
                    if live < 2 || live > 3 {
                        board[r][c] = 2; // live -> dead
                    }
                } else {
                    if live == 3 {
                        board[r][c] = 3; // dead -> live
                    }
                }
            }
        }

        // 2nd pass: finalize
        for r in 0..m {
            for c in 0..n {
                board[r][c] &= 1;
            }
        }
    }
}

fn main() {
    let samples = vec![
        (
            vec![vec![0, 1, 0], vec![0, 0, 1], vec![1, 1, 1], vec![0, 0, 0]],
            vec![vec![0, 0, 0], vec![1, 0, 1], vec![0, 1, 1], vec![0, 1, 0]],
        ),
        (vec![vec![1, 1], vec![1, 0]], vec![vec![1, 1], vec![1, 1]]),
    ];

    for (idx, (input, expected)) in samples.into_iter().enumerate() {
        let mut board = input.clone();
        Solution::game_of_life(&mut board);

        println!("Sample {}: board = {:?}", idx + 1, input);
        println!("Placeholder result: {:?}", board);
        println!("Expected output: {:?}\\n", expected);
    }
}
